### ISSUE TITLE
Testing are failing between test class, but is passing when running individually

### ISSUE THREAD

**Issue by FelipeFMMobile** [kif-framework/KIF#1120]
Some tests are failing (randomly) when running many test classes at once, but when they are running individually they are ok. 

In some tests, I'm trying to put Dispatch main async to hold some tap actions. But is not conclusive, any tips?

**Comment by vrutberg** [comment]
I experienced simular issues when running UI tests in parallel using Xcode 11. Disabling running them in parallel seemed to make them a lot more stable for me. Not running them in parallel is bad, of course, but it could be a temporary solution until there's a real fix.

**Comment by FelipeFMMobile** [comment]
We realise that between tests our alert and views state is not resetting accordingly, keeping it in KIF predicate of identifiers and labels. 

There is a safe way to ensure they clean off in `tearDown` perhaps ?

**Comment by justinseanmartin** [comment]
I don't really have a good answer for you regarding running tests in parallel and if there is something particular that happens in that environment. I don't have enough information to give proper guidance. My understanding was there shouldn't be any shared resources between tests when run this way (e.g. they each get their own simulator), right?

For the general question about app state cleanup, in our tests we always enforce in `afterEach` that  we are in one of two valid states, either on the logged out screen or logged in root screen. In this way, we prevent tests from leaving the app in a state that we can't handle starting a new test from. 

We're in the process of moving to a model where our apps run injected into an empty app host, and then the tests are responsible for setting up the rootViewController. This requires that all app state can be cleanly set back up, but that gives even more isolation between tests. It also means the tests run faster, because we can set the rootViewController to be whatever starting point we want to be testing.

**Comment by vrutberg** [comment]
@justinseanmartin That's exactly what we're in the process of doing. :-)

I'll elaborate some on my comment from earlier and add some recent findings. When Xcode 11 was released we wanted to start running our tests on iOS 13 simulators. Previously, for some reason, they ran on iOS 10 simulators. Mostly the transition to running the tests on iOS 13 was pretty smooth, but we experienced problems where the simulators would just unexplicably freeze when running the tests in parallel. Halting the UI tests at this time and inspecting where it was freezing reveals that it's when KIF uses the clipboard to type text into a textfield. Fast-forward a week or two and we get the same problem with a unit test which also interacts with the clipboard (so the problem is probably not caused by KIF, since this test doesn't use KIF at all and experienced the same problem). Disabling the automatic clipboard sync in Simulator.app has so far seemed to resolve this problem for us, and we have enabled parallel UI testing again.
